plugins/base/src/main/kotlin/renderers/html/NavigationDataProvider.kt

```diff
@@ -0,0 +1,92 @@
+package org.jetbrains.dokka.base.renderers.html
+
+import org.jetbrains.dokka.analysis.PsiDocumentableSource
+import org.jetbrains.dokka.base.renderers.sourceSets
+import org.jetbrains.dokka.base.transformers.documentables.isException
+import org.jetbrains.dokka.model.*
+import org.jetbrains.dokka.pages.*
+import org.jetbrains.dokka.plugability.DokkaContext
+import org.jetbrains.dokka.transformers.pages.PageTransformer
+
+open class NavigationPageInstaller(val context: DokkaContext) : NavigationDataProvider(), PageTransformer {
+
+    override fun invoke(input: RootPageNode): RootPageNode =
+        input.modified(
+            children = input.children + NavigationPage(
+                root = navigableChildren(input),
+                moduleName = context.configuration.moduleName,
+                context = context
+            )
+        )
+}
+
+abstract class NavigationDataProvider {
+    open fun navigableChildren(input: RootPageNode): NavigationNode = input.withDescendants()
+        .first { it is ModulePage || it is MultimoduleRootPage }.let { visit(it as ContentPage) }
+
+    open fun visit(page: ContentPage): NavigationNode =
+        NavigationNode(
+            name = page.displayableName(),
+            dri = page.dri.first(),
+            sourceSets = page.sourceSets(),
+            icon = chooseNavigationIcon(page),
+            children = page.navigableChildren()
+        )
+
+    /**
+     * Parenthesis is applied in 1 case:
+     *  - page only contains functions (therefore documentable from this page is [DFunction])
+     */
+    private fun ContentPage.displayableName(): String =
+        if (this is WithDocumentables && documentables.all { it is DFunction }) {
+            "$name()"
+        } else {
+            name
+        }
+
+    private fun chooseNavigationIcon(contentPage: ContentPage): NavigationNodeIcon? {
+        return if (contentPage is WithDocumentables) {
+            val documentable = contentPage.documentables.firstOrNull()
+            val isJava = (documentable as? WithSources)?.sources?.any { it.value is PsiDocumentableSource } ?: false
```

Comment:
It's a bit ugly and has some excessive operations if you want to use `documentableLanguage()` since it has a `sourceSet` parameter, and we don't have any particular source set here

But I guess avoiding code duplication is more important. Extracted it into a separate function, makes it somewhat better

Reply:
> But I guess avoiding code duplication is more important. Extracted it into a separate function, makes it somewhat better

Yeah, I do not like low level things like `PsiDocumentableSource` (non core API) here. It is not obviously what it means.